contrib/internal/validationtest/validation_test.go

```diff
@@ -0,0 +1,198 @@
+	// augmented with tracing and we don't want these calls to be recorded.
+	// See https://golang.org/pkg/net/http/#DefaultTransport .
+	Transport: &testAgentTransport{
+		Transport: &http.Transport{
+			Proxy:                 http.ProxyFromEnvironment,
+			DialContext:           defaultDialer.DialContext,
+			MaxIdleConns:          100,
+			IdleConnTimeout:       90 * time.Second,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+		},
+	},
+	Timeout: 2 * time.Second,
+}
+
+func TestIntegrations(t *testing.T) {
+	integrations := []Integration{
+		memcachetest.New(),
+		dnstest.New(),
+	}
+	for _, ig := range integrations {
+		name := ig.Name()
+		t.Run(name, func(t *testing.T) {
+			sessionToken := fmt.Sprintf("%s-%d", name, time.Now().Unix())
+			t.Setenv("DD_SERVICE", "Datadog-Test-Agent-Trace-Checks")
+			t.Setenv("DD_TRACE_SPAN_ATTRIBUTE_SCHEMA", "v1")
+			t.Setenv("CI_TEST_AGENT_SESSION_TOKEN", sessionToken)
+
+			tracer.Start(tracer.WithAgentAddr("localhost:9126"), tracer.WithHTTPClient(testAgentClient))
+			defer tracer.Stop()
+
+			cleanup := ig.Init(t)
+			defer cleanup()
+
+			ig.GenSpans(t)
+
+			tracer.Flush()
+
+			assertNumSpans(t, sessionToken, ig.NumSpans())
+			checkFailures(t, sessionToken)
+		})
+	}
+}
+
+func assertNumSpans(t *testing.T, sessionToken string, wantSpans int) {
+	t.Helper()
+	run := func() bool {
+		req, err := http.NewRequest("GET", "http://localhost:9126/test/session/traces", nil)
+		require.NoError(t, err)
+		req.Header.Set("X-Datadog-Test-Session-Token", sessionToken)
+
+		resp, err := http.DefaultClient.Do(req)
+		require.NoError(t, err)
+
+		defer resp.Body.Close()
+		require.Equal(t, http.StatusOK, resp.StatusCode)
+
+		body, err := io.ReadAll(resp.Body)
+		require.NoError(t, err)
+
+		var traces [][]map[string]interface{}
+		require.NoError(t, json.Unmarshal(body, &traces))
+
+		receivedSpans := 0
+		for _, traceSpans := range traces {
+			receivedSpans += len(traceSpans)
+		}
+		if receivedSpans > wantSpans {
+			t.Fatalf("received more spans than expected (wantSpans: %d, receivedSpans: %d)", wantSpans, receivedSpans)
+		}
+		return receivedSpans == wantSpans
+	}
+
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	timeoutChan := time.After(5 * time.Second)
+
+	for {
+		if done := run(); done {
+			return
+		}
+		select {
+		case <-ticker.C:
+			continue
+
+		case <-timeoutChan:
+			t.Fatal("timeout waiting for spans")
+		}
+	}
+}
+
+func checkFailures(t *testing.T, sessionToken string) {
+	t.Helper()
+	req, err := http.NewRequest("GET", "http://localhost:9126/test/trace_check/failures", nil)
+	require.NoError(t, err)
+	req.Header.Set("X-Datadog-Test-Session-Token", sessionToken)
+
+	resp, err := http.DefaultClient.Do(req)
+	require.NoError(t, err)
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
+		return
+	}
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
```

Comment:
These are contradictory statements, please check your code.

Reply:
Lines from 189-191 are an early return as the rest of the function shouldn't run when status == 200.

This line concretely is to display the failure when executing the test. It could be refactored to (I don't remember the if the `bool` returned by `assert.Equal` means success or failure):

```go
if assert.Equal(t, http.StatusOK, resp.StatusCode) {
      return
}
```